components/scream/scripts/gather_all_data.py

```diff
@@ -88,7 +88,7 @@ def formulate_command(self, machine):
         if is_cuda_machine and "OMPI_CXX" not in env_setup_str:
             extra_env = "OMPI_CXX={}/bin/nvcc_wrapper ".format(kokkos_loc)
         else:
-            extra_env = "OMP_PROC_BIND=FALSE KMP_AFFINITY=balanced "
+            extra_env = "OMP_PROC_BIND=FALSE KMP_AFFINITY=balanced OMP_PROC_BIND=spread "
```

Comment:
There's a repetition of OMP_PROC_BIND. If these are processed in this order, it works, but might as well remove the first.

Reply:
Yikes, didn't even notice that. On a related note, OMP_PROC_BIND=spread seems to cause fails on blake:

```
OMP: Warning #181: OMP_PROC_BIND: ignored because KMP_AFFINITY has been defined
OMP: Warning #242: KMP_AFFINITY: granularity=core will be used.
OMP: Warning #238: KMP_AFFINITY: Affinity balanced is not available.
[node22:313804] *** Process received signal ***
[node22:313804] Signal: Floating point exception (8)
[node22:313804] Signal code: Integer divide-by-zero (1)
[node22:313804] Failing at address: 0x7fdfed1faf87
[node22:313804] [ 0] /lib64/libpthread.so.0(+0xf5e0)[0x7fdfecda85e0]
[node22:313804] [ 1] /home/projects/x86-64/intel/compilers/2019/compilers_and_libraries_2019.1.144/linux/compiler/lib/intel64/libiomp5.so(+0x2ff87)[0x7fdfed1faf87]
[node22:313804] [ 2] /home/projects/x86-64/intel/compilers/2019/compilers_and_libraries_2019.1.144/linux/compiler/lib/intel64/libiomp5.so(+0xb5fd2)[0x7fdfed280fd2]
[node22:313804] [ 3] /home/projects/x86-64/intel/compilers/2019/compilers_and_libraries_2019.1.144/linux/compiler/lib/intel64/libiomp5.so(__kmp_fork_call+0x275d)[0x7fdfed27c23d]
[node22:313804] [ 4] /home/projects/x86-64/intel/compilers/2019/compilers_and_libraries_2019.1.144/linux/compiler/lib/intel64/libiomp5.so(__kmpc_fork_call+0x180)[0x7fdfed239e80]
[node22:313804] [ 5] ./p3_tests(_ZN6Kokkos6OpenMP15impl_initializeEi+0xe3)[0x784803]
[node22:313804] [ 6] ./p3_tests[0x7812f5]
[node22:313804] [ 7] ./p3_tests(_ZN6Kokkos10initializeERiPPc+0x1616)[0x780246]
[node22:313804] [ 8] ./p3_tests(_ZN6scream25initialize_scream_sessionEiPPc+0x1f)[0x7722df]
[node22:313804] [ 9] ./p3_tests(main+0x6f)[0x552c3f]
[node22:313804] [10] /lib64/libc.so.6(__libc_start_main+0xf5)[0x7fdfec9f7c05]
[node22:313804] [11] ./p3_tests[0x4bb169]
[node22:313804] *** End of error message ***

78/79 Test #73: p3_run_and_cmp_ut_np1_omp16 ......   Passed    0.56 sec
79/79 Test #72: p3_tests_ut_np1_omp16 ............   Passed    0.89 sec

97% tests passed, 2 tests failed out of 79

Total Test time (real) =  12.03 sec

The following tests FAILED:
         70 - p3_tests_ut_np1_omp14 (NUMERICAL)
         71 - p3_tests_ut_np1_omp15 (NUMERICAL)
```

Reply:
Try just OMP_PROC_BIND=spread, i.e., leave out KMP_AFFINITY. The message "KMP_AFFINITY: Affinity balanced is not available." is new to me and may signal a change in Intel 19 and above. See bottom of this page: https://software.intel.com/en-us/cpp-compiler-developer-guide-and-reference-balanced-affinity-type

Reply:
It says it ignored OMP_PROC_BIND, so it's hard to believe that was the issue. Was this build supposed to have FPE on, or maybe something went wrong when enabling/disabling FPEs?

As for KMP_AFFINITY, it looks overly complicated to set it for a unit test build. We can deal with those shenanigans in a production run, but perhaps for unit tests OMP_PROC_BIND is enough to get "reasonable" times.

Reply:
@ambrad , @bartgol , 

The current value which is only:
```
OMP_PROC_BIND=spread
```

Seems to work well on blake and bowman, but causes significant (5-10x) slowdown on melvin in some heavily threaded tests.

Reply:
Did you double check that melvin's resources are used correctly? If so, then for now maybe just skip the setting on melvin.

Reply:
Double checking now.

Reply:
That was it, melvin must have gotten slammed right around when I ran my test yesterday.

Reply:
OK, I just reconfirmed that melvin performance is good with OMP_PROC_BIND=False. Re-running with spread now.

Reply:
@ambrad , @bartgol , 

I've double-checked and it does appear like melvin performance massively degrades with spread instead of False. I can put logic into gather_all_data to use FALSE instead of spread, but I'd need to know what to check for. Is it use of a non-intel compiler?

Reply:
A combination of reducing CTEST_PARALLEL_LEVEL and newer cmake for melvin has resolve most of the performance loss.